src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+# Code for running https://github.com/EleutherAI/lm-evaluation-harness inside Levanter runs
+# References:
+# https://github.com/kingoflolz/mesh-transformer-jax/blob/master/eval_harness.py
+# https://github.com/kingoflolz/mesh-transformer-jax/blob/f8315e3003033b23f21d78361b288953064e0e76/mesh_transformer/TPU_cluster.py#L6
+import dataclasses
+import functools
+import json
+import logging
+import tempfile
+import typing
+from dataclasses import dataclass
+from functools import cached_property
+from typing import List, Optional, Sequence, Tuple
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+import haliax
+
+import levanter.tracker
+from levanter.compat.hf_checkpoints import HFCheckpointConverter, load_tokenizer
+from levanter.models.gpt2 import Gpt2Config
+from levanter.models.loss import next_token_loss
+
+
+try:
+    from lm_eval import evaluator
+    from lm_eval.api.instance import Instance
+    from lm_eval.api.model import LM
+except ImportError:
+    LM = object
+    Instance = object
+    evaluator = object
+
+from tqdm_loggable.auto import tqdm
+
+import haliax as hax
+from haliax.partitioning import round_axis_for_partitioning
+
+import levanter.config
+from levanter.checkpoint import load_checkpoint
+from levanter.data import AsyncDataset, DataLoader
+from levanter.models.lm_model import LmConfig, LmExample, LmHeadModel
+from levanter.trainer import StepInfo, TrainerConfig
+from levanter.utils.jax_utils import use_cpu_device
+from levanter.utils.tree_utils import inference_mode
+
+
+logger = logging.getLogger(__name__)
+
+
+class EvalDataset(AsyncDataset[LmExample]):
+    def __init__(self, Pos, tokenizer, examples: list[Instance]):
+        super().__init__()
+        self.examples = examples
+        self.Pos = Pos
+        self.tokenizer = tokenizer
+
+    async def async_len(self) -> int:
+        return len(self.examples)
+
+    async def final_length_is_known(self) -> bool:
+        return True
+
+    def is_finite(self) -> bool:
+        return True
+
+    async def current_len(self) -> Optional[int]:
+        return len(self.examples)
+
+    async def get_batch(self, indices: Sequence[int]) -> List[LmExample]:
+        out = []
+        pad_token_id = self.tokenizer.pad_token_id
+
+        reqs = [(self.examples[i].args[0], self.examples[i].args[1]) for i in indices]
```

Comment:
Comment what assumptions are made about the structure of the examples (args[0], args[1])?